// @trusted-pdf/pdf/src/syntax/structure/trailer.ts
import { Names } from "../names";
import { PdfArray, PdfDictionary, PdfNumber } from "../objects";
import { Catalog } from "./catalog";

export class Trailer extends PdfDictionary {
  public get Size(): number {
    return this.get(Names.Size, PdfNumber).value;
  }
  public set Size(value: number) {
    this.get(Names.Size, PdfNumber).value = value;
  }

  public get Prev(): number | null {
    return this.find(Names.Prev, PdfNumber)?.value ?? null;
  }
  public set Prev(value: number | null) {
    if (value) {
      this.set(Names.Prev, new PdfNumber({ value }));
    }
  }

  public get Root(): Catalog {
    return this.get(Names.Root, Catalog);
  }
  public set Root(value: PdfDictionary) {
    this.set(Names.Root, value);
  }

  public get ID(): PdfArray | null {
    return this.find(Names.ID, PdfArray);
  }
  public set ID(value: PdfArray | null) {
    if (value) {
      this.set(Names.ID, value);
    }
  }

  public get Info(): PdfDictionary | null {
    return this.find(Names.Info, PdfDictionary);
  }
  public set Info(value: PdfDictionary | null) {
    if (value) {
      this.set(Names.Info, value);
    }
  }

  public get Encrypt(): PdfDictionary | null {
    return this.find(Names.Encrypt, PdfDictionary);
  }
  public set Encrypt(value: PdfDictionary | null) {
    if (value) {
      this.set(Names.Encrypt, value);
    }
  }

  public override onCreate(): void {
    this.set(Names.Size, this.doc.createNumber(1));
  }
}
